Updated bulk importer to improve performance

allouis opened the pull request:
no-issue

The main changes here are:
- Refactor to remove global state, hopefully easier to maintain
- Do everything we can do in parallel
- Use a Leaky Bucket to make requests to Stripe API as fast as possible
- Do not add members that we know we will delete (e.g. ones with stripe
  customers when stripe is not enabled)
- Remove as many loops as possible, e.g. combine filter and map into a
  single reduce
- Handle live & test stripe rate limits
- Remove support for Complimentary plans (it has to be reworked anyway)

allouis commented:
@naz I've used a PR for this & requested your review on this because it's quite a big changeset, and I didn't wanna push to master and screw you over.

Basic idea here is:
1. Do as many things in parallel as we can
2. Use the Leaky Bucket to request as fast as possible from Stripe API

naz commented on core/server/api/canary/members.js:
```diff
@@ -604,7 +604,10 @@ module.exports = {
                 });
-            }).then(() => {
+            }).then((result) => {
+                invalid.errors = invalid.errors.concat(result.invalid.errors);
```
Yesss! +100 points for getting rid of those 2 extra parameters!

allouis replied:
:hugs:

naz commented on core/server/services/members/importer/index.js:
```diff
@@ -80,150 +179,92 @@ const doImport = async ({members, allLabelModels, importSetLabels, imported, inv
-            }
-        };
+async function fetchStripeCustomers(stripeCustomersToInsert) {
+    const LeakyBucket = require('leaky-bucket');
```
Awesome one. Definitely a thing to rely upon stead of testing at which concurrency count things break :+1:

allouis replied:
Yeah! The implementation is pretty big though, has a lot of unnecessary deps :confused: The newest version looks cool, but needs to be run with an experimental flag :weary:  Might be worth writing our own, or finding a better library in the future.

naz commented on core/server/services/members/importer/index.js:
```diff
@@ -4,50 +4,153 @@ const ObjectId = require('bson-objectid');
-    if (_.isString(labels)) {
-        if (labels === '') {
-            return [];
+const CHUNK_SIZE = 5000;
```
Would leave a note here about the CHUNK_SIZE value - we still need to test what's the optimal value for it is. Based on [these RoTs](http://mysql.rjweb.org/doc.php/ricksrots) the number should be between 100 and 1000.

naz commented on core/server/services/members/importer/index.js:
```diff
@@ -4,50 +4,153 @@ const ObjectId = require('bson-objectid');
+            return;
+        }
+
+        // @TODO This is expensive, maybe we can just error if we get shoddy data?
```
I'd even be up for dropping this whole block because it was added to handle a bug that happened in members export [pre version 3.6](https://github.com/TryGhost/Ghost/blob/78911257f407da311cf6a62bc148a51e5de4a4fb/core/server/api/canary/members.js#L16-L19) - limited time window. This is just my opinion, would need verification!

allouis replied:
Okay, maybe one for @matthanley to take a look at?

TL;DR pre 3.6 we would export the string 'undefined' for undefined values, and we now have to loop through every key in every object to handle that - which is expensive, especially for large imports.

Would you consider dropping the support for that?  We could provide a command/tool to clean the data up locally if necessary OR we could offload that work to Ghost-Admin, that way the operation would not affect the request cycle

naz commented:
@allouis I do like the direction and think we should merge these once the import works for for all needed scenarios. The missing bits I have in mind for these to be mergeable: members_stripe_customers_subscriptions records insertion and the `comped` flag handling. This is an awesome improvement!

naz commented:
@allouis should we merge this into master as is? Asking because that would help fixing the bug with rate-limit error in "validate import" :wink:

allouis committed:
Added bulkAdd method to Member model

no-issue

This allows us to keep the db access in the model layer

allouis committed:
Updated @tryghost/members-api to 0.27.2

no-issue

This includes fixes for rate-limiting of requests, and exposes necessary
Stripe methods for creating customers and complimentary subscriptions,
without affecting the database.

allouis committed:
Refactored importer to parallelise tasks where possible

no-issue

By parallelising our tasks we are able to improve the speed at which the
entire import completes.

allouis committed:
Implemented bulkAdd for Customer & Subscription models

no-issue
